File: Checkers.cs
// Checkers.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Checkers
{
    const int SIZE = 8;
    const int EMPTY = 0;
    const int WHITE = 1;
    const int BLACK = 2;
    const int WHITE_KING = 3;
    const int BLACK_KING = 4;

    int[,] board = new int[SIZE, SIZE];
    int turn = WHITE;
    Random rand = new Random();

    public Checkers()
    {
        InitBoard();
    }

    void InitBoard()
    {
        for (int r = 0; r < 3; r++)
            for (int c = 0; c < SIZE; c++)
                if ((r + c) % 2 == 1) board[r, c] = BLACK;
        for (int r = 5; r < 8; r++)
            for (int c = 0; c < SIZE; c++)
                if ((r + c) % 2 == 1) board[r, c] = WHITE;
    }

    bool IsWhite(int p) => p == WHITE || p == WHITE_KING;
    bool IsBlack(int p) => p == BLACK || p == BLACK_KING;
    bool IsKing(int p) => p == WHITE_KING || p == BLACK_KING;
    int Opponent(int p) => IsWhite(p) ? BLACK : WHITE;

    bool InBounds(int r, int c) => r >= 0 && r < SIZE && c >= 0 && c < SIZE;
    int GetPiece(int r, int c) => InBounds(r, c) ? board[r, c] : EMPTY;

    bool IsEnemy(int r, int c, int piece)
    {
        int p = GetPiece(r, c);
        if (p == EMPTY) return false;
        if (IsWhite(piece) && IsBlack(p)) return true;
        if (IsBlack(piece) && IsWhite(p)) return true;
        return false;
    }

    bool CanCapture(int r, int c, int dr, int dc)
    {
        int nr = r + dr, nc = c + dc;
        if (!InBounds(nr, nc)) return false;
        if (GetPiece(nr, nc) == EMPTY) return false;
        if (!IsEnemy(nr, nc, board[r, c])) return false;
        int lr = r + 2*dr, lc = c + 2*dc;
        if (!InBounds(lr, lc)) return false;
        if (GetPiece(lr, lc) != EMPTY) return false;
        return true;
    }

    List<(int, int)> CaptureMovesFrom(int r, int c)
    {
        int piece = board[r, c];
        var dirs = new[] { (-1,-1), (-1,1), (1,-1), (1,1) };
        if (!IsKing(piece))
        {
            if (IsWhite(piece)) dirs = new[] { (-1,-1), (-1,1) };
            else dirs = new[] { (1,-1), (1,1) };
        }
        var moves = new List<(int, int)>();
        foreach (var (dr, dc) in dirs)
            if (CanCapture(r, c, dr, dc))
                moves.Add((r + 2*dr, c + 2*dc));
        return moves;
    }

    List<(int, int)> SimpleMovesFrom(int r, int c)
    {
        int piece = board[r, c];
        var dirs = new[] { (-1,-1), (-1,1), (1,-1), (1,1) };
        if (!IsKing(piece))
        {
            if (IsWhite(piece)) dirs = new[] { (-1,-1), (-1,1) };
            else dirs = new[] { (1,-1), (1,1) };
        }
        var moves = new List<(int, int)>();
        foreach (var (dr, dc) in dirs)
        {
            int nr = r + dr, nc = c + dc;
            if (InBounds(nr, nc) && GetPiece(nr, nc) == EMPTY)
                moves.Add((nr, nc));
        }
        return moves;
    }

    bool HasCaptures(int color)
    {
        for (int r = 0; r < SIZE; r++)
            for (int c = 0; c < SIZE; c++)
            {
                int p = board[r, c];
                if (p == EMPTY) continue;
                if (color == WHITE && !IsWhite(p)) continue;
                if (color == BLACK && !IsBlack(p)) continue;
                if (CaptureMovesFrom(r, c).Count > 0) return true;
            }
        return false;
    }

    List<(int, int)> LegalMovesFrom(int r, int c)
    {
        int piece = board[r, c];
        if (piece == EMPTY) return new List<(int, int)>();
        if (IsWhite(piece) && turn != WHITE) return new List<(int, int)>();
        if (IsBlack(piece) && turn != BLACK) return new List<(int, int)>();
        if (HasCaptures(turn))
            return CaptureMovesFrom(r, c);
        return SimpleMovesFrom(r, c);
    }

    bool ApplyMove(int fr, int fc, int tr, int tc)
    {
        int piece = board[fr, fc];
        if (piece == EMPTY) return false;
        int dr = tr - fr, dc = tc - fc;
        if (Math.Abs(dr) == 2 && Math.Abs(dc) == 2)
        {
            int cr = fr + dr/2, cc = fc + dc/2;
            if (GetPiece(cr, cc) == EMPTY) return false;
            board[cr, cc] = EMPTY;
        }
        board[tr, tc] = piece;
        board[fr, fc] = EMPTY;
        if (!IsKing(piece))
        {
            if (IsWhite(piece) && tr == 0) board[tr, tc] = WHITE_KING;
            else if (IsBlack(piece) && tr == SIZE-1) board[tr, tc] = BLACK_KING;
        }
        turn = Opponent(piece);
        return true;
    }

    int Winner()
    {
        int white = 0, black = 0;
        for (int r = 0; r < SIZE; r++)
            for (int c = 0; c < SIZE; c++)
            {
                int p = board[r, c];
                if (IsWhite(p)) white++;
                else if (IsBlack(p)) black++;
            }
        if (white == 0) return BLACK;
        if (black == 0) return WHITE;
        return EMPTY;
    }

    void Display()
    {
        Console.Clear();
        Console.WriteLine("  a b c d e f g h");
        for (int r = 0; r < SIZE; r++)
        {
            Console.Write($"{r+1} ");
            for (int c = 0; c < SIZE; c++)
            {
                int p = board[r, c];
                char ch;
                switch (p)
                {
                    case EMPTY: ch = '.'; break;
                    case WHITE: ch = 'w'; break;
                    case BLACK: ch = 'b'; break;
                    case WHITE_KING: ch = 'W'; break;
                    case BLACK_KING: ch = 'B'; break;
                    default: ch = '?'; break;
                }
                Console.Write($"{ch} ");
            }
            Console.WriteLine($"{r+1}");
        }
        Console.WriteLine("  a b c d e f g h");
        Console.WriteLine($"Turn: {(turn == WHITE ? "White" : "Black")}");
    }

    bool PlayerMove()
    {
        while (true)
        {
            Display();
            string color = turn == WHITE ? "White" : "Black";
            Console.Write($"{color}, enter move (e.g., a3-b4) or 'q' to quit: ");
            string input = Console.ReadLine().Trim();
            if (input == "q") return false;
            if (input == "u") { Console.WriteLine("Undo not implemented."); continue; }
            var parsed = ParseMove(input);
            if (parsed == null) { Console.WriteLine("Invalid format."); continue; }
            var (fr, fc, tr, tc) = parsed.Value;
            int piece = board[fr, fc];
            if (piece == EMPTY) { Console.WriteLine("No piece there."); continue; }
            if ((IsWhite(piece) && turn != WHITE) || (IsBlack(piece) && turn != BLACK))
            { Console.WriteLine("Not your turn."); continue; }
            var legal = LegalMovesFrom(fr, fc);
            if (!legal.Any(m => m.Item1 == tr && m.Item2 == tc))
            { Console.WriteLine("Illegal move."); continue; }
            if (!ApplyMove(fr, fc, tr, tc))
            { Console.WriteLine("Move failed."); continue; }
            int w = Winner();
            if (w != EMPTY)
            {
                Display();
                Console.WriteLine(w == WHITE ? "White wins!" : "Black wins!");
                return false;
            }
            return true;
        }
    }

    (int fr, int fc, int tr, int tc)? ParseMove(string str)
    {
        var parts = str.Split('-');
        if (parts.Length != 2) return null;
        string from = parts[0].Trim(), to = parts[1].Trim();
        if (from.Length != 2 || to.Length != 2) return null;
        int fc = from[0] - 'a';
        int fr = int.Parse(from[1].ToString()) - 1;
        int tc = to[0] - 'a';
        int tr = int.Parse(to[1].ToString()) - 1;
        if (fr < 0 || fr >= SIZE || fc < 0 || fc >= SIZE || tr < 0 || tr >= SIZE || tc < 0 || tc >= SIZE)
            return null;
        return (fr, fc, tr, tc);
    }

    bool AIMove()
    {
        var moves = new List<(int fr, int fc, int tr, int tc)>();
        for (int r = 0; r < SIZE; r++)
            for (int c = 0; c < SIZE; c++)
            {
                int p = board[r, c];
                if (p == EMPTY) continue;
                if ((IsWhite(p) && turn != WHITE) || (IsBlack(p) && turn != BLACK)) continue;
                var legal = LegalMovesFrom(r, c);
                foreach (var (tr, tc) in legal)
                    moves.Add((r, c, tr, tc));
            }
        if (moves.Count == 0) { Console.WriteLine("AI has no moves."); return false; }
        var move = moves[rand.Next(moves.Count)];
        ApplyMove(move.fr, move.fc, move.tr, move.tc);
        int w = Winner();
        if (w != EMPTY)
        {
            Display();
            Console.WriteLine(w == WHITE ? "White wins!" : "Black wins!");
            return false;
        }
        return true;
    }

    static void Main()
    {
        var game = new Checkers();
        Console.WriteLine("Russian Checkers");
        Console.WriteLine("Controls: enter moves as 'a3-b4'. 'q' to quit.");
        while (true)
        {
            if (game.turn == WHITE)
            {
                if (!game.PlayerMove()) break;
            }
            else
            {
                if (!game.AIMove()) break;
            }
        }
    }
}

File: checkers.go
// checkers.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const SIZE = 8
const (
	EMPTY = iota
	WHITE
	BLACK
	WHITE_KING
	BLACK_KING
)

type Checkers struct {
	board      [SIZE][SIZE]int
	turn       int
	moveHistory []string
}

func NewCheckers() *Checkers {
	c := &Checkers{turn: WHITE}
	c.initBoard()
	return c
}

func (c *Checkers) initBoard() {
	for r := 0; r < 3; r++ {
		for col := 0; col < SIZE; col++ {
			if (r+col)%2 == 1 {
				c.board[r][col] = BLACK
			}
		}
	}
	for r := 5; r < 8; r++ {
		for col := 0; col < SIZE; col++ {
			if (r+col)%2 == 1 {
				c.board[r][col] = WHITE
			}
		}
	}
}

func (c *Checkers) isWhite(p int) bool { return p == WHITE || p == WHITE_KING }
func (c *Checkers) isBlack(p int) bool { return p == BLACK || p == BLACK_KING }
func (c *Checkers) isKing(p int) bool  { return p == WHITE_KING || p == BLACK_KING }
func (c *Checkers) opponent(p int) int {
	if c.isWhite(p) {
		return BLACK
	}
	return WHITE
}

func (c *Checkers) inBounds(r, c int) bool {
	return r >= 0 && r < SIZE && c >= 0 && c < SIZE
}

func (c *Checkers) getPiece(r, c int) int {
	if c.inBounds(r, c) {
		return c.board[r][c]
	}
	return EMPTY
}

func (c *Checkers) isEnemy(r, c int, piece int) bool {
	p := c.getPiece(r, c)
	if p == EMPTY {
		return false
	}
	if c.isWhite(piece) && c.isBlack(p) {
		return true
	}
	if c.isBlack(piece) && c.isWhite(p) {
		return true
	}
	return false
}

func (c *Checkers) canCapture(r, c, dr, dc int) bool {
	nr, nc := r+dr, c+dc
	if !c.inBounds(nr, nc) {
		return false
	}
	if c.getPiece(nr, nc) == EMPTY {
		return false
	}
	if !c.isEnemy(nr, nc, c.board[r][c]) {
		return false
	}
	lr, lc := r+2*dr, c+2*dc
	if !c.inBounds(lr, lc) {
		return false
	}
	if c.getPiece(lr, lc) != EMPTY {
		return false
	}
	return true
}

func (c *Checkers) captureMovesFrom(r, c int) [][2]int {
	var moves [][2]int
	piece := c.board[r][c]
	dirs := [][2]int{{-1, -1}, {-1, 1}, {1, -1}, {1, 1}}
	// For kings, all directions; for others, only forward
	if !c.isKing(piece) {
		if c.isWhite(piece) {
			// white moves up (row decreases)
			dirs = [][2]int{{-1, -1}, {-1, 1}}
		} else {
			// black moves down (row increases)
			dirs = [][2]int{{1, -1}, {1, 1}}
		}
	}
	for _, d := range dirs {
		dr, dc := d[0], d[1]
		if c.canCapture(r, c, dr, dc) {
			lr, lc := r+2*dr, c+2*dc
			moves = append(moves, [2]int{lr, lc})
		}
	}
	return moves
}

func (c *Checkers) simpleMovesFrom(r, c int) [][2]int {
	var moves [][2]int
	piece := c.board[r][c]
	dirs := [][2]int{{-1, -1}, {-1, 1}, {1, -1}, {1, 1}}
	if !c.isKing(piece) {
		if c.isWhite(piece) {
			dirs = [][2]int{{-1, -1}, {-1, 1}}
		} else {
			dirs = [][2]int{{1, -1}, {1, 1}}
		}
	}
	for _, d := range dirs {
		nr, nc := r+d[0], c+d[1]
		if c.inBounds(nr, nc) && c.getPiece(nr, nc) == EMPTY {
			moves = append(moves, [2]int{nr, nc})
		}
	}
	return moves
}

func (c *Checkers) hasCaptures(color int) bool {
	for r := 0; r < SIZE; r++ {
		for c := 0; c < SIZE; c++ {
			p := c.board[r][c]
			if p == EMPTY {
				continue
			}
			if color == WHITE && !c.isWhite(p) {
				continue
			}
			if color == BLACK && !c.isBlack(p) {
				continue
			}
			if len(c.captureMovesFrom(r, c)) > 0 {
				return true
			}
		}
	}
	return false
}

func (c *Checkers) legalMovesFrom(r, c int) [][2]int {
	piece := c.board[r][c]
	if piece == EMPTY {
		return nil
	}
	if c.isWhite(piece) && c.turn != WHITE {
		return nil
	}
	if c.isBlack(piece) && c.turn != BLACK {
		return nil
	}
	if c.hasCaptures(c.turn) {
		return c.captureMovesFrom(r, c)
	}
	return c.simpleMovesFrom(r, c)
}

func (c *Checkers) applyMove(fr, fc, tr, tc int) bool {
	piece := c.board[fr][fc]
	if piece == EMPTY {
		return false
	}
	dr, dc := tr-fr, tc-fc
	if abs(dr) == 2 && abs(dc) == 2 {
		cr, cc := fr+dr/2, fc+dc/2
		if c.getPiece(cr, cc) == EMPTY {
			return false
		}
		c.board[cr][cc] = EMPTY
	}
	c.board[tr][tc] = piece
	c.board[fr][fc] = EMPTY
	// Promotion
	if !c.isKing(piece) {
		if c.isWhite(piece) && tr == 0 {
			c.board[tr][tc] = WHITE_KING
		} else if c.isBlack(piece) && tr == SIZE-1 {
			c.board[tr][tc] = BLACK_KING
		}
	}
	// Switch turn
	if c.turn == WHITE {
		c.turn = BLACK
	} else {
		c.turn = WHITE
	}
	return true
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func (c *Checkers) winner() int {
	white, black := 0, 0
	for r := 0; r < SIZE; r++ {
		for c := 0; c < SIZE; c++ {
			p := c.board[r][c]
			if c.isWhite(p) {
				white++
			} else if c.isBlack(p) {
				black++
			}
		}
	}
	if white == 0 {
		return BLACK
	}
	if black == 0 {
		return WHITE
	}
	return EMPTY
}

func (c *Checkers) display() {
	fmt.Print("\033[H\033[2J")
	fmt.Println("  a b c d e f g h")
	for r := 0; r < SIZE; r++ {
		fmt.Printf("%d ", r+1)
		for c := 0; c < SIZE; c++ {
			p := c.board[r][c]
			var ch string
			switch p {
			case EMPTY:
				ch = "."
			case WHITE:
				ch = "w"
			case BLACK:
				ch = "b"
			case WHITE_KING:
				ch = "W"
			case BLACK_KING:
				ch = "B"
			}
			fmt.Printf("%s ", ch)
		}
		fmt.Printf("%d\n", r+1)
	}
	fmt.Println("  a b c d e f g h")
	if c.turn == WHITE {
		fmt.Println("Turn: White")
	} else {
		fmt.Println("Turn: Black")
	}
}

func parseMove(moveStr string) (int, int, int, int, error) {
	parts := strings.Split(moveStr, "-")
	if len(parts) != 2 {
		return 0, 0, 0, 0, fmt.Errorf("invalid format")
	}
	from := parts[0]
	to := parts[1]
	if len(from) != 2 || len(to) != 2 {
		return 0, 0, 0, 0, fmt.Errorf("invalid square")
	}
	fc := int(from[0] - 'a')
	fr, _ := strconv.Atoi(string(from[1]))
	fr--
	tc := int(to[0] - 'a')
	tr, _ := strconv.Atoi(string(to[1]))
	tr--
	if fr < 0 || fr >= SIZE || fc < 0 || fc >= SIZE || tr < 0 || tr >= SIZE || tc < 0 || tc >= SIZE {
		return 0, 0, 0, 0, fmt.Errorf("out of bounds")
	}
	return fr, fc, tr, tc, nil
}

func (c *Checkers) playerMove() bool {
	reader := bufio.NewReader(os.Stdin)
	for {
		c.display()
		color := "White"
		if c.turn == BLACK {
			color = "Black"
		}
		fmt.Printf("%s, enter move (e.g., a3-b4) or 'q' to quit: ", color)
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)
		if input == "q" {
			return false
		}
		if input == "u" {
			fmt.Println("Undo not implemented.")
			continue
		}
		fr, fc, tr, tc, err := parseMove(input)
		if err != nil {
			fmt.Println("Invalid format:", err)
			continue
		}
		piece := c.board[fr][fc]
		if piece == EMPTY {
			fmt.Println("No piece there.")
			continue
		}
		if c.isWhite(piece) && c.turn != WHITE {
			fmt.Println("Not your turn.")
			continue
		}
		if c.isBlack(piece) && c.turn != BLACK {
			fmt.Println("Not your turn.")
			continue
		}
		legal := c.legalMovesFrom(fr, fc)
		found := false
		for _, m := range legal {
			if m[0] == tr && m[1] == tc {
				found = true
				break
			}
		}
		if !found {
			fmt.Println("Illegal move.")
			continue
		}
		if !c.applyMove(fr, fc, tr, tc) {
			fmt.Println("Move failed.")
			continue
		}
		if w := c.winner(); w != EMPTY {
			c.display()
			if w == WHITE {
				fmt.Println("White wins!")
			} else {
				fmt.Println("Black wins!")
			}
			return false
		}
		return true
	}
}

func (c *Checkers) aiMove() bool {
	var moves [][4]int // fr, fc, tr, tc
	for r := 0; r < SIZE; r++ {
		for c := 0; c < SIZE; c++ {
			p := c.board[r][c]
			if p == EMPTY {
				continue
			}
			if c.isWhite(p) && c.turn != WHITE {
				continue
			}
			if c.isBlack(p) && c.turn != BLACK {
				continue
			}
			legal := c.legalMovesFrom(r, c)
			for _, m := range legal {
				moves = append(moves, [4]int{r, c, m[0], m[1]})
			}
		}
	}
	if len(moves) == 0 {
		fmt.Println("AI has no moves.")
		return false
	}
	// Random move
	move := moves[rand.Intn(len(moves))]
	fr, fc, tr, tc := move[0], move[1], move[2], move[3]
	c.applyMove(fr, fc, tr, tc)
	if w := c.winner(); w != EMPTY {
		c.display()
		if w == WHITE {
			fmt.Println("White wins!")
		} else {
			fmt.Println("Black wins!")
		}
		return false
	}
	return true
}

func main() {
	rand.Seed(time.Now().UnixNano())
	game := NewCheckers()
	fmt.Println("Russian Checkers")
	fmt.Println("Controls: enter moves as 'a3-b4'. 'q' to quit.")
	for {
		if game.turn == WHITE {
			if !game.playerMove() {
				break
			}
		} else {
			if !game.aiMove() {
				break
			}
		}
	}
}
